### PR TITLE
v2

### PR DESCRIPTION
**Pull request by prabirshrestha**
This v2 is a complete rewrite of asyncomplete while still maintaining 100% api compatibility with sources api. Some features have been removed and moved to user land while some of the irritating bugs have been fixed.

Here are some notable changes:
- [x] added support g:asyncomplete_popup_delay so the popup shows after a dealy
- [x] add concept of asyncomplete change manager. This is in an internal api but allows clear separation to notify changes on text. One is implemented using `TextChangedP` and the other is implemented with timers if `TextChangedP` is not supported.
- [x] Add support for trigger characters similar to that of language server protocol which are faster since it is O(1) and easier to write compared to regex.
- [x] Force refresh
- [x] avoid duplicate normalization of completion items
- [x] allow users to provide custom preprocessor that modifies the completion items. This can be used to support fuzzy search, remove duplicates in the user land.
- [x] do not force refresh when pressing backspace chars. This was a bug in v1 and pressing backspace could trigger completor func to be called causing serious lag when using with vim-lsp on big files.
- [x] Do not throw away results by slow completors. In v1 if you were using slow completor and you typed other chars, context_changed() would return true causing the results to be thrown away. This would then automatically cause a completor() request and if that is also slow it would be thrown away and doing this several keystorkes makes vim slow if you are using a heavy completor. v2 now caches it instead of throwing away making it very fast. 
- [x] add option to not override `completeopt`.
- [x] remove lua and smart completion support. Use `g:asyncomplete_preprocessor` instead. I get too much request to modify the result completed items. This is now a function so it can be achieved in the user space.
- [x] update docs to reflect changes
- [x] update README.md to reflect changes
- [x] update asyncomplete-lsp and others sources to support trigger characters. https://github.com/prabirshrestha/asyncomplete.vim/pull/124
- [x] add support for `complete_info()` so can ignore when omni/file and other native vim completions are used. Waiting for https://github.com/vim/vim/pull/3866, https://github.com/vim/vim/pull/4106 and https://github.com/neovim/neovim/pull/9815 to be merged.

Post v2 features that is currently out of scope for this PR but I would be more than happy to include in this PR if they send PR against v2 branch.
- [ ] Add support for multiple character trigger characters that are common in some languages such as c++. `::` and `->`
- [ ] Use trigger characters to detect word separation instead of regex. I need to think more about this. Not sure if it is a good or a bad idea.
- [ ] add support for min char length - https://github.com/prabirshrestha/asyncomplete.vim/pull/122
- [ ] allow async support for `g:asyncomplete_preprocessor`. This enables interesting ideas such as using python threads to implement fuzzy search without blocking vim, most recently used (MRU), machine learning 😄 , enable grouping of candidates - https://github.com/prabirshrestha/asyncomplete.vim/pull/103. This is one of the core reason why `g:asyncomplete_smart_completion` has been removed. core will only contain prefix search. 
- [ ] fix bug where sometimes pressing `o` or `O` might cause the popup to auto trigger even if no chars is entered.

Please give this a try and report bugs. Hopefully can merge this in a week. Feel free to ask any questions.

### CODE REVIEW

File: doc/asyncomplete.txt [hunk 1]
```diff
@@ -14,80 +14,76 @@ CONTENTS                                                *asyncomplete-contents*
...
 
-g:asyncomplete_auto_completeopt                       *g:asyncomplete_auto_completeopt*
+      function! s:my_asyncomplete_preprocessor(ctx, matches) abort
+        let l:dict = {}
+        let l:items = []
+        for [l:source_name, l:matches] in items(a:matches)
+          for l:item in l:matches['items']
+            if l:item['word'] =~ '^' . a:options['base']
+              if !has_key(l:dict, l:item['word'])
+                call add(l:items, l:item)
```
Comment (machakann) [inline doc/asyncomplete.txt]: Probably, `let l:dict[l:item['word']] = 1` or something is required here?

Comment (prabirshrestha) [reply doc/asyncomplete.txt]: fixed.

**Comment by mattn** [comment]
path completion <kbd>CTRL</kbd>-<kbd>x</kbd> <kbd>CTRL</kbd>-<kbd>f</kbd> open/close popup immediately. Sometines works correctly.

terminal7

**Comment by prabirshrestha** [comment]
@mattn That also happens for `ctrl-x ctrl-o` omni-func and yes I do find it irritating. It is to avoid calling the filter func multiple times. For example if you have two or more sync function such as buffer and syntax as soon it buffer calls `asyncomplete#complete` it will filter and then again once syntax calls `asyncomplete#complete` it will filter again. currently it only filters once.

https://github.com/prabirshrestha/asyncomplete.vim/blob/416723c5353bb9c688e7a5993c12f9e974508888/autoload/asyncomplete.vim#L329-L336 Instead of queuing with timer can you try calling `s:recompute_pum` all the time in `update_pum` method and see if that helps.

Ideally I have always wanted vim to provide us an api that tells us that pum was opened by  `ctrl-x ctrl-o` or `ctrl-x ctrl-y` so I can ignore on_change for asyncomplete. If vim doesn't support this then we might need to have a config for now to disable optimize filtering.

**Comment by prabirshrestha** [comment]
Seems like it won't work even if I call `s:recompute_pum()` directly without timers in `s:update_pum()` as typing next character will cause the filter function to execute and will ignore the `ctrl-x ctrl-f` 😢 Seems like we need an api in vim to support this. Not sure how other complete engines solves this.

**Comment by mattn** [comment]
What the API for vim?

**Comment by prabirshrestha** [comment]
My thought was to have something like this.

```vim
autocmd CompleteOpen * call s:complete_open()
autocmd CompleteDone * call s:complete_done()

function! s:complete_open() abort
    let s:complete_type = v:event['complete_type']
endfunction

function! s:complete_done() abort
    let s:complete_type = ''
endfunction

function! s:on_changed() abort
    if s:complete_type != 'complete' | return | endif
    " .... asyncomplete s:on_change code goes here
endfunction
```

Basically I need to know have an event when the pum was opened and if the pum was opened by asyncomplete or someone else such as `ctrl-x ctrl-f`. If the pum is opened by someone else I need to ignore. 

I'm trying out if we can figure this out by just `CompleteDone` and without adding any patches to vim.

**Comment by prabirshrestha** [comment]
Actually don't think it can be done by just `CompleteDone`. Can you try this patch. This partially solves the issue but means we can't have a custom preprocessor now.

```vim
diff --git a/autoload/asyncomplete.vim b/autoload/asyncomplete.vim
index 3a4ccb0..c9f373e 100644
--- a/autoload/asyncomplete.vim
+++ b/autoload/asyncomplete.vim
@@ -377,7 +377,9 @@ function! s:recompute_pum(...) abort
         \ }, l:ctx)

     if empty(g:asyncomplete_preprocessor)
-        call s:default_preprocessor(l:filter_ctx, l:matches_to_filter)
+        if !pumvisible()
+            call s:default_preprocessor(l:filter_ctx, l:matches_to_filter)
+        endif
     else
         call g:asyncomplete_preprocessor[0](l:filter_ctx, l:matches_to_filter)
     endif
```

If we have slow sources it will not work since the pum will be visible. I will need to most likely add a flag when a `asyncomplete#complete()` is called always force show the pum even if it has already changed.

**Comment by mattn** [comment]
The patch fixes my issue.

File: autoload/asyncomplete.vim [hunk 2]
```diff
@@ -10,141 +16,87 @@ if !has('timers')
...
 
-function! asyncomplete#force_refresh() abort
-    return asyncomplete#menu_selected() ? "\<c-y>\<c-r>=asyncomplete#_force_refresh()\<CR>" : "\<c-r>=asyncomplete#_force_refresh()\<CR>"
-endfunction
-
-function! asyncomplete#_force_refresh() abort
-    if get(b:, 'asyncomplete_enable')
-        call s:remote_refresh(asyncomplete#context(), 1)
+function! asyncomplete#unregister_source(info_or_server_name) abort
+    if type(a:info) == type({})
```
Comment (mattn) [inline autoload/asyncomplete.vim]: This func dose not have argument a:info.

Comment (prabirshrestha) [reply autoload/asyncomplete.vim]: Fixed.

File: autoload/asyncomplete.vim [hunk 3]
```diff
@@ -10,141 +16,87 @@ if !has('timers')
...
-        for l:event in a:info['events']
-            let l:exec =  'if get(b:,"asyncomplete_enable",0) | call s:python_cm_event("' . a:info['name'] . '", "'.l:event.'",asyncomplete#context()) | endif'
-            if type(l:event) == type('')
-                execute 'au ' . l:event . ' * ' . l:exec
-            elseif type(l:event) == type([])
-                execute 'au ' . join(l:event,' ') .' ' .  l:exec
-            endif
-        endfor
-        execute 'augroup end'
+        doautocmd User asyncomplete_setup
```
Comment (mattn) [inline autoload/asyncomplete.vim]: Please define this User event on the top of this file.

```vim
augroup _asyncomplete_silent_
    autocmd!
    autocmd User asyncomplete_setup silent
augroup END
```

Comment (prabirshrestha) [reply autoload/asyncomplete.vim]: Fixed. Thanks.

**Comment by prabirshrestha** [comment]
Currently blocked to merging on master due to https://github.com/vim/vim/pull/3866

**Comment by prabirshrestha** [comment]
@mattn Here is the patch for making omni/file completion work with asyncomplete v2. Still waiting for https://github.com/vim/vim/pull/3866 to be merged to vim and neovim.

```diff
diff --git a/autoload/asyncomplete.vim b/autoload/asyncomplete.vim
index 8bf426b..c24d29f 100644
--- a/autoload/asyncomplete.vim
+++ b/autoload/asyncomplete.vim
@@ -16,6 +16,7 @@ if !has('timers')
     finish
 endif

+let s:has_complete_info = exists('*complete_info')
 let s:already_setup = 0
 let s:sources = {}
 let s:matches = {} " { server_name: { incomplete: 1, startcol: 0, items: [], refresh: 0, status: 'idle|pending|success|failure', ctx: ctx } }
@@ -382,12 +383,22 @@ function! s:recompute_pum(...) abort
         \ 'startcol': l:startcol,
         \ }, l:ctx)

-    if empty(g:asyncomplete_preprocessor)
-        if !pumvisible()
-            call s:default_preprocessor(l:filter_ctx, l:matches_to_filter)
+    let l:ignore_preprocessor = 0
+    if s:has_complete_info
+        let l:complete_info = complete_info()
+        if !empty(l:complete_info) && has_key(l:complete_info, 'mode')
+            let l:ignore_preprocessor = !(l:complete_info['mode'] == '' || l:complete_info['mode'] == 'eval')
+        endif
+    endif
+
+    if !l:ignore_preprocessor
+        if empty(g:asyncomplete_preprocessor)
+            if !pumvisible()
+                call s:default_preprocessor(l:filter_ctx, l:matches_to_filter)
+            endif
+        else
+            call g:asyncomplete_preprocessor[0](l:filter_ctx, l:matches_to_filter)
         endif
-    else
-        call g:asyncomplete_preprocessor[0](l:filter_ctx, l:matches_to_filter)
     endif
 endfunction
```

File: autoload/asyncomplete.vim [hunk 4]
```diff
@@ -357,178 +163,263 @@ function! s:get_active_sources_for_buffer() abort
...
 
-function! s:custom_filter_completion_items(prefix, matches) abort
-    let l:start = s:is_nvim ? 1 : 0
-    let l:last = s:is_nvim ? len(a:matches) : len(a:matches) - 1
-    return luaeval('asyncomplete.filter_completion_items(_A.prefix, _A.matches, _A.start, _A.last)', { 'prefix': a:prefix, 'matches': a:matches, 'start': l:start, 'last': l:last })
+function! s:default_preprocessor(options, matches) abort
+    let l:items = []
+    for [l:source_name, l:matches] in items(a:matches)
+        for l:item in l:matches['items']
+            if l:item['word'] =~ '^' . a:options['base']
```
Comment (mattn) [inline autoload/asyncomplete.vim]: This must be:
```
if stridx(l:item['word'], a:options['base']) == 0
```
Typing `~` goes fail.

Comment (prabirshrestha) [reply autoload/asyncomplete.vim]: Fixed. Thanks.

**Comment by prabirshrestha** [comment]
This is now merged since both vim and neovim now support `complete_info()`. Thanks for the feedback.

v1 has been tagged with v1.7 and v2 is now tagged with v2.0.0 and v2 branch has now been deleted.

**Comment by prabirshrestha** [comment]
Here is an example of custom preprocessor. Seems very interesting. https://github.com/machakann/asyncomplete-ezfilter.vim

**Pull request merged**